### DISCUSSION TITLE
Is there a collision detection method for meshes?

### DISCUSSION THREAD

**Discussion by Boehm92** [jimy-byerley/pymadcad#38] [Q&A] [answered]
Hi, 

i want to difference multiple machining features from a cube randomly. For example, i want to difference a 10mm and 5mm drill hole from the cube. If i substract first the 10mm drill hole and then the 5mm at the same place, there will be an exception because there is no mesh i can difference the 5mm from. 

So, i would like to randomly place the 10mm and 5mm drillholes first and check if both colide. If so, i unite both with the boolean operation and then difference them from the cube. 

**Therefore, i need a method to check, if both drillholes are colliding after randomly placing them. Is there such a function?**

Kind Regards,
Stefan

**Comment by jimy-byerley** [comment, marked as answer]
Hello @Boehm92 !

## About drilling

I'm not sure to understand the need of an intersection-check function in this case, as there is nothing wrong in performing a boolean union or difference when the meshes are not intersecting. In case of non intersecting meshes, a union will simply concatenate them, resulting in one mesh containing 2 *islands*, a difference will result in only the first mesh unmodified.
Here is an example of what I understand of your case:
```python
from madcad import *

def cylinder(bottom:vec3, top:vec3, radius:float) -> 'Mesh':
	direction = top-bottom
	return extrusion(direction, 
             flatsurface(wire(Circle((bottom,normalize(direction)), radius)))
             .flip())

cube = brick(width=vec3(1))
c1 = cylinder(O, Z, 0.2)
c2 = cylinder(vec3(1,0,0.1), vec3(1,0,0.9), 0.2)
c3 = cylinder(vec3(0.2,0,0.2), vec3(0.2,0,0.8), 0.15)

# both following solutions are valid:
result = difference(difference(difference(cube, c1), c2), c3)     # performing drill holes one by one
result = difference(cube, union(union(c1, c2), c3))      # making a union of the drills and then performing the holes all at once
```
- The cylinders
  ![image](https://user-images.githubusercontent.com/7940753/166108136-bb60ddbf-934f-44fb-ae07-b3dae2253372.png)

- The result
  ![image](https://user-images.githubusercontent.com/7940753/166108144-e5927974-c653-43fa-9981-7b15fecf65f4.png)

## About intersection-check

For the question of intersection-check itself, I might have a solution, but not very efficient ... The lowest level intersection function available in the API is [`madcad.boolean.cut_mesh()`](https://pymadcad.readthedocs.io/en/latest/reference/boolean.html#madcad.boolean.cut_mesh).  boolean operation on meshes are built on top of this function.

However it is meant to *actually* compute intersections, so it will compute all the intersections before returning, instead of stopping at the first existing intersection found, as a good intersection-check would do.
Also it is an in-place operation (altering the input mesh), so not suited for checks.

If you really need to check for intersections (or possible intersections) between meshes, I would advice creating dedicated functions, using for instance the following techniques:
- bounding-box intersections check  (to check for possible intersections)
- convex hull intersection check  (to check more precisely for possible intersections)
- search for intersections using [spacial hashing](https://pymadcad.readthedocs.io/en/latest/reference/hashing.html#madcad.hashing.PositionMap)  (to quickly access potentially intersecting faces)
- return at the first intersection found 
- eventually use [pybullet](https://github.com/babbatem/pybullet_collision/blob/master/example.py) to check in a vast amount of meshes

If you really need such a function I think I could hack something :thinking:

**Reply by Boehm92** [reply]
Hi @jimy-byerley ,

i tried your code your code and also my own code with those two SolidWork-.STL models [SolidWorks Models.zip](https://github.com/jimy-byerley/pymadcad/files/8601618/SolidWorks.Models.zip) , then i get following exception:

```python
cube = mdc.read('CadTemplates/Cube.stl')
cube.mergeclose()

cylinder = mdc.read('cylinder.stl')
cylinder.mergeclose()

first_time_diff_cylinder = mdc.difference(cube, cylinder)

cylinder = cylinder.transform(mdc.mat3(0.9, 0.9, 0.9))
cylinder = cylinder.transform(mdc.vec3(-3, -3, 0))

second_time_diff_cylinder = mdc.difference(first_time_diff_cylinder, cylinder)
```

![Diff_Example1](https://user-images.githubusercontent.com/40231115/166205475-734f6c11-0255-4ab8-8e13-2144d02c2779.PNG)


>   File "C:\Users\...\Desktop\Repos\MachiningFeatureRecognizer\venv\lib\site-packages\madcad\boolean.py", line 242, in boolean_mesh
>     if not selector[0] and selector[1]:mc2 = mc2.flip()
> AttributeError: 'set' object has no attribute 'flip'


If i import the cylinder as a new object "cylinder2", i get this exception:

```python
cube = mdc.read('CadTemplates/Cube.stl')
cube.mergeclose()

cylinder = mdc.read('cylinder.stl')
cylinder.mergeclose()

first_time_diff_cylinder = mdc.difference(cube, cylinder)

cylinder2 = mdc.read('cylinder.stl')
cylinder2 = cylinder2.transform(mdc.mat3(0.9, 0.9, 0.9))
cylinder2 = cylinder2.transform(mdc.vec3(-3, -3, 0))
cylinder2.mergeclose()

second_time_diff_cylinder = mdc.difference(first_time_diff_cylinder, cylinder)
```

> File "C:\Users\..\Desktop\Repos\MachiningFeatureRecognizer\venv\lib\site-packages\madcad\triangulation.py", line 284, in triangulation_outline
>     raise TriangulationError("no more feasible triangles (algorithm failure or bad input outline)", [outline.indices[i] for i in hole])


Do you have an idea what the problem could be?


Kind Regards
Stefan

**Reply by jimy-byerley** [reply]
I'm unable to reproduce the same `AttributeError` :thinking:  what is strange is that the variable `mc2` should not be a `set` at the point raising, there must be something wrong higher in the complete backtrace ...

For the second exception, this seems to be the same issue as #8 
Your cylinder top and bottom surfaces are coincident with your cube's faces, which raises local ambiguities for the algorithm.
Currently the only workaround is to add a small offset to avoid overlapping surfaces
which gives you the following script
```python
import madcad as mdc

cube = mdc.read('/tmp/SolidWorks.Models/Cube.stl')
cube.mergeclose()
cube = segmentation(cube)  # segmentation is for aspect, when importing STL files

cylinder = mdc.read('/tmp/SolidWorks.Models/cylinder.stl')
cylinder.mergeclose()
cylinder = mdc.segmentation(cylinder.transform(vec3(1e-3)))  # segmentation for aspect, and small offset between surfaces

first_time_diff_cylinder = mdc.difference(cube, cylinder)

cylinder = cylinder.transform(mdc.mat3(0.9, 0.9, 0.9))

second_time_diff_cylinder = mdc.difference(first_time_diff_cylinder, cylinder)
```

**Reply by Boehm92** [reply]
Thanks again for the fast respons. I'm greatfull for your great support!

In your example the cylinder is positioned in the cube the second time. There, i also dont get the "AttributeError". But when i move the cylinder outside the cube (in empty space),  just then i get the "Attribute Error".

```python
from madcad import *

cube = read('Cube.stl')
cube.mergeclose()
cube = segmentation(cube)  # segmentation is for aspect, when importing STL files

cylinder = read('cylinder.stl')
cylinder.mergeclose()
cylinder = segmentation(cylinder.transform(vec3(1e-3)))  # segmentation for aspect, and small offset between surfaces

first_time_diff_cylinder = difference(cube, cylinder)

cylinder = cylinder.transform(mat3(0.9, 0.9, 0.9))
cylinder = cylinder.transform(vec3(-3, -3, 0))

show([cube, cylinder, cylinder])

second_time_diff_cylinder = difference(first_time_diff_cylinder, cylinder)

**Reply by jimy-byerley** [reply]
Hi @Boehm92 , well you might have an old version of the module don't you ?
Event when I reinstall the pure v0.11 from pypi I'm unable to reproduce the `AttributeError`
what's the result of `pip show pymadcad` ?

**Reply by jimy-byerley** [reply]
If `pip` doesn't show you `pymadcad == 0.11` I would suggest upgrading it using `pip install -U pymadcad`

**Reply by Boehm92** [reply]
That did the trick. I allready tried to update the package, but for some reason it stuck by 0.10.2. I created a new interpreter and it worked.

Thanks so much and sorry for all the trouble!

**Reply by jimy-byerley** [reply]
No problem, you are welcome :) I'm glad to help
